### ISSUE TITLE
Improve API Documentation   (apidocs)

### ISSUE THREAD

**Issue by crookedneighbor** [HabitRPG/habitica#8087]
_EDIT by Alys 2018-06-30:_
 
The only remaining action for this PR is to document the member route. It's been started (probably mostly done) by @madpink in https://github.com/HabitRPG/habitica/pull/9760 so that needs to be taken over by someone and finished (see the staff comments in that PR).

----

The docs for the API routes are, for the most part, very sparse. Often they don't provide enough information for actually using them.

There are a bunch of parameters that we can use with [API Doc](http://apidocjs.com/#params).

A good example of how each route should look is the [invite to group route](https://habitica.com/apidoc/#api-Group-InviteToGroup). Here it is in the [source code](https://github.com/HabitRPG/habitrpg/blob/ee15e29ba4d34812c4fd4e5836dc310d713a7db9/website/server/controllers/api-v3/groups.js#L643-L683).

When opening a PR, you should do it a little at a time. NOT ALL AT ONCE. For instance, one PR for each section.
- [x] Challenge -  @MathWhiz -- _(a mostly complete solution by JTorr is at https://github.com/HabitRPG/habitica/pull/8112 and if MathWhiz uses it in a PR that is merged, JTorr will get partial credit towards a tier)_  https://github.com/HabitRPG/habitica/pull/8626
- [x] Chat - @MathWhiz https://github.com/HabitRPG/habitica/pull/8277
- [x] Content (just document the top level parts of the object) - @MathWhiz #8098
- [x] Coupon - @MathWhiz https://github.com/HabitRPG/habitica/pull/8109
- [x] Data Export - @MathWhiz https://github.com/HabitRPG/habitica/pull/8120
- [x] Development - @MathWhiz https://github.com/HabitRPG/habitica/pull/8480
- [x] Group - @MathWhiz https://github.com/HabitRPG/habitica/pull/8522
  - [x] Invite to Group - @JTorr #8091
- [x] Hall - @Althenis #8536
- [ ] Member - continue from https://github.com/HabitRPG/habitica/pull/9760
- [x] Meta - @MathWhiz https://github.com/HabitRPG/habitica/pull/8167
- [x] Status - @MathWhiz #8165 
- [x] Tag - @madpink #8400 
- [x] Task - @madpink https://github.com/HabitRPG/habitica/pull/8447
- [x] Unsubscribe - @MathWhiz #8187 
- [x] User - a maximum of 10 routes per PR:
   - [x] [GET `/api/v3/user`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L26) – [POST `/api/v3/user/buy/:key`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L560) - @madpink #8476
   - [x] [POST `/api/v3/user/buy-gear/:key`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L580) – [POST `/api/v3/user/change-class`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L785) @madpink https://github.com/HabitRPG/habitica/pull/8602
   - [x] [POST `/api/v3/user/disable-classes`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L810) – [POST `/api/v3/user/unlock`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L1007) @madpink https://github.com/HabitRPG/habitica/pull/8720
   - [x] [POST `/api/v3/user/revive`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L1031) – [POST `/api/v3/user/custom-day-start`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L1267) @madpink https://github.com/HabitRPG/habitica/pull/8792

**Comment by JTorr** [comment]
I'm interested in tackling this.

**Comment by crookedneighbor** [comment]
How about you claim a specific group, so if someone else wants to work on another section, they can.

**Comment by crookedneighbor** [comment]
Section.

**Comment by JTorr** [comment]
Sounds good. I'll work on the Challenge section.

**Comment by JTorr** [comment]
I'm submitting a pull request to revise the docs for the Invite to Group section. I had just completed a pull request in that section of code, and noticed a few things that needed updating. I'll be moving on to the Challenges section next.

**Comment by librarianmage** [comment]
I can work on Content and Coupon

**Comment by crookedneighbor** [comment]
For the content route, just document the top level parts of the object.

**Comment by richardhooper** [comment]
Could I have a crack at Tags?

**Comment by Alys** [comment]
@richardhooper That would be great, thanks! 

Everyone, I've updated the top post with the claims so far.

**Comment by librarianmage** [comment]
What info should I put into the Coupons section? I'm not sure what a request example would be.

**Comment by librarianmage** [comment]
How do I generate the apiDoc client side?

**Comment by paglias** [comment]
With `gulp apidoc` or `gulp apidoc:watch`

**Comment by librarianmage** [comment]
Should we list `NotAuthorized` errors? If so, what message should we use to keep things consistent?

**Comment by librarianmage** [comment]
I can work on data export

**Comment by librarianmage** [comment]
For the [@apiParam](http://apidocjs.com/#param-api-param) option, it allows parameters to be grouped. Should we group them into (say) body parameters, query parameters, and path?

**Comment by crookedneighbor** [comment]
If that works, we should.

**Comment by librarianmage** [comment]
@crookedneighbor I looked into it, and I would recommend putting these snippets at the top of each file:

Body Parameters:

``` js
/**
 * @apiDefine bodyParam Body Parameters
 */
```

Query Parameters:

``` js
/**
 * @apiDefine queryParam Query Parameters
 */
```

Path Parameters:

``` js
/**
 * @apiDefine pathParam Path Parameters
 */
```

(Descriptions can also be added if necessary)

Then, `(:groupName:)` can be inserted after the `@apiParam` instruction.

**Comment by crookedneighbor** [comment]
They only need to be put in one file. Anything you define in one is accessible everywhere.

**Comment by crookedneighbor** [comment]
I've added them https://github.com/HabitRPG/habitica/commit/afb5b473a37325a82151bf48b0c7d54b4979e6eb

**Comment by khipkin** [comment]
From @crookedneighbor's comment [here](https://github.com/HabitRPG/habitica/pull/8109/files#r82929833):

> Errors should be in the form:
> `* apiError (STATUS_CODE) {ErrorType} SpecificErrorName Description about the error`

**Comment by librarianmage** [comment]
I can do meta, status, and unsubscribe!

**Comment by akortesmaa** [comment]
I can do Hall & Member

**Comment by librarianmage** [comment]
I can do chat!

Other Updates:
- Coupon is finished
- Status is finished ( #8165 )
- Unsubscribe is finished ( #8187 )

**Comment by madpink** [comment]
I would like to help with tasks

**Comment by Alys** [comment]
@madpink Thank you!

**Comment by librarianmage** [comment]
@Althenis Are you still planning on working on Hall and Member?

**Comment by madpink** [comment]
I've just made one change thus far, but before I wade too deeply into it, I wanted to make sure that I am on the right path. There really are a lot of potential errors, because any body param could have produce an error.

I include some examples, but that really increases the shear amount of space being used... are they a good idea?

**Comment by akortesmaa** [comment]
@MathWhiz First of all, thanks for the reminder. Hall seems to be almost done, so I'll try to finalize it in the next couple of days. Haven't done anything on member though, so feel free to take it if you want.

**Comment by madpink** [comment]
Tasks is finished,  waiting on approval.

I would be happy to work on more.

May I take over tags?

**Comment by richardhooper** [comment]
@madpink Be my guest, I'm afraid I never got up to steam!

**Comment by MathWhiz** [comment]
Note: Meta has been merged in

**Comment by librarianmage** [comment]
* Meta has been finished
* Chat is located at #8277

**Comment by madpink** [comment]
please change the PR reference on tasks from 8400 to 8447

**Comment by Alys** [comment]
@madpink Done! Thanks for the note!

**Comment by madpink** [comment]
quick question about APIDOC, are apiuse definitions shared from document or do they need to be defined each time?

**Comment by librarianmage** [comment]
From what I remember, they don't require re-definition

**Comment by librarianmage** [comment]
I can do Development!

**Comment by madpink** [comment]
Tag and Task have both been merged, I am going to start wading into "user" has anyone already started on it?

**Comment by librarianmage** [comment]
@madpink From the comments, it appears that no one has started working on it, so you should have no problems starting!

@Alys @crookedneighbor I took a peek at the current state of the api documentation, and would suggest this break-down for PRs (keeping in mind the 10 max routes/PR suggested):

* [GET `/api/v3/user`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L26) – [POST `/api/v3/user/buy/:key`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L560)
* [POST `/api/v3/user/buy-gear/:key`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L580) – [POST `/api/v3/user/change-class`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L785)
* [POST `/api/v3/user/disable-classes`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L810) – [POST `/api/v3/user/unlock`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L1007)
* [POST `/api/v3/user/revive`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L1031) – [POST `/api/v3/user/custom-day-start`](https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/user.js#L1267)

**Comment by Alys** [comment]
I've broken down `user` as suggested by @MathWhiz 

@madpink I've assigned you to the first set but let us known if you've started on one of the others.

**Comment by madpink** [comment]
okay dokeey... I will begin on the first segment

here's a quick question, I like including sample output, but for GET /api/v3/user even with a new user it is extremely long (almost 600 lines)... which would make the most sense:
-abbreviating the output by removing some lines
-skipping the output altogether when it is long
-including the whole thing
-move output to a wiki page and link back

**Comment by paglias** [comment]
@madpink I think abbreviating the output in places where there are long lists (for example items, email preferences, ...) is the way to go. Also because the attributes in those lists change quite frequently and we don't want to update them every time an item is added

**Comment by madpink** [comment]
I am having a problem with the delete user command. I create a dummy account to delete, but I can't get it right. 

I tried the following curl command:

`curl -X DELETE -H “Content-Type:application/json” -H “x-api-user:xxxxxxxx-5a12-4a06-91e6-6877b6b6ea6a” -H “x-api-key:xxxxxxxx-a225-4e5b-b178-14bb9064eed4” -d {“password”:”abc123”} https://habitica.com/api/v3/user`

and I tried:
`curl -X DELETE -H “Content-Type:application/json” -H “x-api-user:xxxxx-5a12-4a06-91e6-6877b6b6ea6a” -H “x-api-key:xxxxxxxxxx-a225-4e5b-b178-14bb9064eed4” https://habitica.com/api/v3/user?password=abc123`

I keep getting:
`{"success":false,"error":"NotAuthorized","message":"There is no account that uses those credentials."}`

I've double checked the UserID and API key several times... can anyone spot my mistake?

**Comment by Alys** [comment]
@madpink I can confirm that the UID and Token are correct. :) You left them in place in the second curl command, but I've edited them out since your test account has an email address in it that might be a real one. If it weren't a test account that was going to be deleted, I'd change the token. 

Does `DELETE` work like `POST` or `PUT`? I can't recall. If so, the syntax for a body parameter is more like  this:
`curl -X PUT -d "text=Edited Task Text" -H "x-api-user: 12345678-....`
from http://habitica.wikia.com/wiki/Application_Programming_Interface
So you'd put password=... where that example has text=.... and change PUT to DELETE of course.

When you're updating the docs for that route, can you change "The user's password (unless it's a Facebook account)" to "The user's password if the account uses local authentication" ?

**Comment by madpink** [comment]
@Alys Thanks, and oops.  I've tried numerous methods to send the delete user command, I still cannot get any result other than either "There is no account that uses those credentials." or "Invalid request parameters."

Sending a payload through a DELETE statement is fairly unusual, so is there possibly another step I am missing?

Trying to trace my way through the script, it seems as though I'm passing the checkBody which seems to tell me that the body is being received, and the password is being recognized as present, does it need to be encoded in some way?

**Comment by akortesmaa** [comment]
@madpink `curl -X DELETE -H "x-api-user: ***-44a8-b98e-ec4468a7e467" -H "x-api-key: ***-4550-431c-8f4d-4769bc25ec25" -d password=1234 localhost:3000/api/v3/user` worked for me. Response is `{"success":true,"data":{},"notifications":[]}`

**Comment by madpink** [comment]
`curl -X DELETE -H “x-api-user:xxxxxxx-5a12-4a06-91e6-6877b6b6ea6a” -H “x-api-key:xxxxxxxx-a225-4e5b-b178-14bb9064eed4” -d password=abc123 https://habitica.com/api/v3/user`

Still will not work for me, I am 100% sure the userID, password and apiKey are correct, but I still keep getting
`{"success":false,"error":"NotAuthorized","message":"There is no account that uses those credentials."}`

**Comment by akortesmaa** [comment]
@madpink So you are absolutely sure that the account still exists? Can you `GET` its profile? Do you still get the same response if you create another account and try to delete that?

**Comment by paglias** [comment]
If it keeps happening can you send me the ID so I can check if everything
is fine?

Il 04 Feb 2017 19:51, "Atte Kortesmaa" <notifications@github.com> ha
scritto:

**Comment by madpink** [comment]
Thanks for the responses... I was able to execute a number of actions on this account such as allocating points, adding tasks and changing class, but oddly I could not DELETE the account nor could I GET the user profile. I have since created another fake account and was able to delete it.

@paglias I sent you the account info if you want to tinker with it.

**Comment by madpink** [comment]
I put in a PR for the first batch of User, will start working on the next

**Comment by madpink** [comment]
I have the second batch of User ready and will submit soon (waiting for merger of part 1)
I am going to move on to third batch

**Comment by librarianmage** [comment]
Can this be re-opened?

**Comment by TheHollidayInn** [comment]
Ah, this closed because you had "Fixes". I would not add that if you are only doing a partial fix.

**Comment by librarianmage** [comment]
Yep! I totally forgot to remove it when submitting my PR

**Comment by madpink** [comment]
in anyone familiar with how user/unlock is supposed to work?
I tried the following POST's, both of which appeared to work, took gems out of the account, but did not actually do anything

1.  "https://habitica.com/api/v3/user/unlock?path=midnight_clouds"
2 gems were deducted from my account, but Midnight Clouds did not unlock. (and it should have cost 7 gems)

2. "https://habitica.com/api/v3/user/unlock?path=spookySkins"
same thing, 2 gems were deducted, but nothing was unlocked in my account (and it should have cost 5 gems)

**Comment by Alys** [comment]
@madpink From a quick look at the code here: https://github.com/HabitRPG/habitica/blob/develop/website/common/script/ops/unlock.js#L12-L32
it looks like the path you specify needs to be a full path (e.g., containing the word "background" for your first example), and it's costing 2 gems (the default 0.5 cost) and failing because you've given it a partial path.

For a background, the path in the user's document is `purchased.background.midnight_clouds` but it might be referring to a path in the `content` data.

This probably indicates that we need to put more error checking on that route. :scream_cat: We shouldn't be taking gems when the user / third-party-dev uses an incorrect path. @madpink Would you like to create a new issue for that?

Did you test in production and hence lose real gems? If so, drop your User ID into a comment here and I'll check it out and do a refund.

**Comment by madpink** [comment]
thanks @Alys I did test in production, UUID: b0413351-405f-416f-8787-947ec1c85199

for the record, I am almost done with the entire "User" section... I have 5 calls left to test, and these I definitely need to use in a test environment

will do some more testing and then submit an issue

**Comment by Alys** [comment]
@madpink I've given you four gems.

**Comment by madpink** [comment]
I have finished the User section, I am waiting on the first PR to be merged before submitting the next.

I would be willing to pick up on Challenge if needed, or any other sections that might need help

**Comment by librarianmage** [comment]
I can work on Group!

**Comment by librarianmage** [comment]
For Groups, should I document how `create-plan` works?

**Comment by librarianmage** [comment]
@Althenis How is Hall going?

**Comment by librarianmage** [comment]
@JTorr Are you still working on Challenge?

**Comment by Alys** [comment]
@MathWhiz @JTorr I'm opening up Challenges for someone else to pick up. For more details see my comment at https://github.com/HabitRPG/habitica/pull/8112#issuecomment-285779308 . I'm updating the top post to link to the PR so it can be reused but to say that Challenges are open. 

@MathWhiz If you want to take on Challenges, post here and I'll assign it to you. Otherwise it's open for anyone.

**Comment by librarianmage** [comment]
@Alys I will take Challenges on

**Comment by librarianmage** [comment]
It appears to have been missed in the main message, but #8536 is for the Hall.

**Comment by librarianmage** [comment]
Challenge docs can be checked off!

**Comment by Alys** [comment]
As reported by MrTyx (cbcf8442-3be2-4f36-9331-ba277555c08b) in the Aspiring Comrades guild: 

- [x] https://habitica.com/apidoc/#api-Quest-InviteToQuest Route is listed as https://habitica.com/api/v3/groups/:groupId/quests/invite but should be https://habitica.com/api/v3/groups/:groupId/quests/invite/:questKey - @madpink  #8720

Since the groups routes have already been done, someone could add this to a PR they're doing for other routes. Please comment here to say when you've done that so that we don't get duplicated PRs.

EDIT: this is being done by @madpink

**Comment by madpink** [comment]
Added questKey update to PR #8720

**Comment by akortesmaa** [comment]
PR #8536 Should be ready now.

**Comment by madpink** [comment]
at this point, are there any others that need work?

**Comment by Alys** [comment]
The Member routes are the only remaining ones. They were assigned to Althenis who then said [above](https://github.com/HabitRPG/habitica/issues/8087#issuecomment-272304071) that MathWhiz could do it, although we didn't change the assignment in the top post. 

MathWhiz are you working on Member?

**Comment by librarianmage** [comment]
Oops, not yet. @madpink can complete it if they like.

**Comment by madpink** [comment]
I'd be happy to work on Member

**Comment by Alys** [comment]
@madpink It's yours, thanks.

**Comment by madpink** [comment]
I have Member pretty much ready except for:
 /api/v3/members/:toUserId/objections/:interaction

How do I actually initiate an objection in order to then view it with this command?

**Comment by librarianmage** [comment]
@madpink How are you doing on this?

**Comment by madpink** [comment]
@MathWhiz It's pretty much ready... I need to rebase and check for changes then I'll submit it.

**Comment by librarianmage** [comment]
@madpink Hello! How are you doing with the Member section?

**Comment by Alys** [comment]
Another problem to be fixed in website/server/controllers/api-v3/tasks.js  :
`@api {get} /api/v3/tasks/user` can take a `dueDate` parameter but that's not mentioned in the apidoc comments  for that route. It should be.

**Comment by Alys** [comment]
For the `dueDate` parameter mentioned in my previous comment, I've added placeholder documentation in this commit: https://github.com/HabitRPG/habitica/commit/ade14edcd70fb1b17f8fd00d2302039c1eadf42b

What's the purpose of that parameter? When you specify it, it doesn't affect which Dailies (or other tasks) are returned. Instead, it changes the `nextDue` dates in Dailies in the returned data:
https://github.com/HabitRPG/habitica/blob/ade14edcd70fb1b17f8fd00d2302039c1eadf42b/website/server/libs/taskManager.js#L211-L215
https://github.com/HabitRPG/habitica/blob/ade14edcd70fb1b17f8fd00d2302039c1eadf42b/website/server/libs/taskManager.js#L25-L55

 It doesn't save those changes to the database.

What's the use case for that behaviour?

**Comment by beffymaroo** [comment]
@Alys Is this issue still in process? Should any label changes be made since the associated PR is abandoned?

**Comment by Alys** [comment]
I've marked it help wanted but it probably most needs to have the remaining items broken into new issues (or just handled rapidly if the fixes are small). If no one's claimed this by the weekend, I'll try to make time to clean it up.

I'd still need information about the `dueDate` parameter from my comment just above. Maybe from @TheHollidayInn ?

**Comment by TheHollidayInn** [comment]
The duedate is for the mobile apps. Basically it allows for the `nextDue` field to be computed based on a specific date. 

So, starting from yesterday, I can see the next few due dates. Or starting from last week.

**Comment by Alys** [comment]
The only remaining action for this PR is to document the member route. It's been started (probably mostly done) by @madpink in https://github.com/HabitRPG/habitica/pull/9760 so that needs to be taken over by someone and finished (see the staff comments in that PR). I've updated the top post to include this summary.

**Comment by JacobFrericks** [comment]
Any objections if I take this?

**Comment by paglias** [comment]
Go ahead @JacobFrericks !

**Comment by TheHollidayInn** [comment]
I think there is only one PR left linked above if you want to finish that

**Comment by ChariseWalraven** [comment]
Does #9760 still need to be done?

**Comment by JacobFrericks** [comment]
@ChariseWalraven I've taken over on a separate PR (since #9760 has been abandoned). It's just waiting for approval.
